### PR TITLE
Add a cache for initialSync responses that expires after 5 minutes

### PR DESCRIPTION
**Pull request by NegativeMjark**
Needs  matrix-org/sytest#124 to land first

### CODE REVIEW

File: synapse/util/caches/snapshot_cache.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            self.next_result_cache = {}
+            self.time_last_rotated_ms = time_now_ms
+
+    def get(self, time_now_ms, key):
+        self.rotate(time_now_ms)
+        # This cache is intended to deduplicate requests, so we expect it to be
+        # missed most of the time. So we just lookup the key in all of the
+        # dictionaries rather than trying to short circuit the lookup if the
+        # key is found.
+        result = self.prev_result_cache.get(key)
```
Comment (leonerd) [inline synapse/util/caches/snapshot_cache.py]: ITYM `.get(key, None)` ?

Comment (leonerd) [inline synapse/util/caches/snapshot_cache.py]: Uh nevermind... Though maybe put that as default anyway to not confuse the next person? ;)

File: synapse/util/caches/snapshot_cache.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        result = self.next_result_cache.get(key, result)
+        result = self.pending_result_cache.get(key, result)
+        if result is not None:
+            return result.observe()
```
Comment (leonerd) [inline synapse/util/caches/snapshot_cache.py]: Maybe add a fallthrough "return None" to make explicit what happens on a cache miss?

File: synapse/util/caches/snapshot_cache.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from synapse.util.async import ObservableDeferred
+
+
+class SnapshotCache(object):
```
Comment (leonerd) [inline synapse/util/caches/snapshot_cache.py]: Wants a small docstring to explain what it does. Maybe a small description of the different semantics of pending, next and prev dicts, as that is slightly subtle.

File: synapse/handlers/message.py [hunk 4]
```diff
@@ -326,9 +328,29 @@ def get_state_events(self, user_id, room_id, is_guest=False):
...
+            return result
+
+        return self.snapshot_cache.set(now_ms, key, self._snapshot_all_rooms(
+            user_id, pagin_config, as_client_event, include_archived
+        ))
+
+    @defer.inlineCallbacks
+    def _snapshot_all_rooms(self, user_id=None, pagin_config=None,
+                            as_client_event=True, include_archived=False):
         """Retrieve a snapshot of all rooms the user is invited or has joined.
```
Comment (leonerd) [inline synapse/handlers/message.py]: Move the docstring up to the "public facing" non-_-prefixed method

File: synapse/util/caches/snapshot_cache.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/util/caches/snapshot_cache.py]: Done d12c00b

File: synapse/util/caches/snapshot_cache.py (hunk 3 above)
Comment (NegativeMjark) [inline synapse/util/caches/snapshot_cache.py]: Done d12c00b

**Comment by richvdh** [comment]
lgtm

**Pull request merged**